MapboxNavigation/VanishingRouteLine.swift

```diff
@@ -0,0 +1,372 @@
+    */
+    public func updateUpcomingRoutePointIndex(routeProgress: RouteProgress) {
+        guard let completeRoutePoints = primaryRoutePoints else {
+            primaryRouteRemainingDistancesIndex = nil
+            return
+        }
+        let currentLegProgress = routeProgress.currentLegProgress
+        let currentStepProgress = routeProgress.currentLegProgress.currentStepProgress
+        /**
+        Find the count of remaining points in the current step.
+        */
+        var allRemainingPoints = getSlicedLinePointsCount(currentLegProgress: currentLegProgress, currentStepProgress: currentStepProgress)
+        
+        /**
+         Add to the count of remaining points all of the remaining points on the current leg, after the current step.
+        */
+        let currentLegSteps = completeRoutePoints.nestedList[routeProgress.legIndex]
+        let startIndex = currentLegProgress.stepIndex + 1
+        let endIndex = currentLegSteps.count - 1
+        if startIndex < endIndex {
+            allRemainingPoints += currentLegSteps.prefix(endIndex).suffix(from: startIndex).flatMap{ $0.compactMap{ $0 } }.count
+        }
+        
+        /**
+         Add to the count of remaining points all of the remaining legs.
+        */
+        for index in stride(from: routeProgress.legIndex + 1, to: completeRoutePoints.nestedList.count, by: 1) {
+            allRemainingPoints += completeRoutePoints.nestedList[index].flatMap{ $0 }.count
+        }
+        
+        /**
+         After calculating the number of remaining points and the number of all points,  calculate the index of the upcoming point.
+        */
+        let allPoints = completeRoutePoints.flatList.count
+        primaryRouteRemainingDistancesIndex = allPoints - allRemainingPoints - 1
+    }
+    
+    func getSlicedLinePointsCount(currentLegProgress: RouteLegProgress, currentStepProgress: RouteStepProgress) -> Int {
+        let startDistance = currentStepProgress.distanceTraveled
+        let stopDistance = currentStepProgress.step.distance
+        var slicedLine: Turf.LineString?
+        
+        /**
+         Implement the Turf.lineSliceAlong(lineString, startDistance, stopDistance) to return a sliced lineString.
+         */
+        if let lineString = currentStepProgress.step.shape {
+            if let midPoint = lineString.coordinateFromStart(distance: startDistance) {
+                slicedLine = lineString.trimmed(from: midPoint, distance: stopDistance - startDistance)
+            }
+        }
+        
+        if slicedLine != nil {
+            return slicedLine!.coordinates.count - 1
+        } else {
+            return 0
+        }
+    }
+    
+    func calculateRouteGranularDistances(coordinates: [CLLocationCoordinate2D]) -> RouteLineGranularDistances? {
+        if coordinates.isEmpty {
+            return nil
+        } else {
+            return calculateGranularDistances(coordinates)
+        }
```

Comment:
What about more traditional:
```suggestion
        if !coordinates.isEmpty {
            return calculateGranularDistances(coordinates)
        }
        
        return nil
```

Reply:
Thanks for the suggestion, I'll fix it

Reply:
Thanks for the suggestion. At the start of the navigation, because the `fractionTraveled` hasn't be updated, there's no available lineGradient to update. I think it can be solved by a fixer to use the `routeProgress.fractionTraveled` instead or use speed to update the lineGradient. I'll work on this problem to fix it.

Reply:
The refresh problem and the start of the navigation problem are both caused by the `navigationMapView.show(_ routes: [Route], legIndex: Int = 0)`. inside the function, it defines the `mainRouteLayer` and `parentLayer` with a `lineGradient` that forms from a 0.0 fractionTraveled. 

So at the start of the navigation, before it receives the routeProgress to call `func updateRoute(_ routeProgress:)`, it shows the vanishing route line with the stop at the start point.  It causes the gap between the puck view and not perfectly synced.

When refresh, the `func navigationService(_ service: NavigationService, didRefresh routeProgress: RouteProgress)` also call the `navigationMapView.show(_ routes: [Route], legIndex: Int = 0)` first to make the vanishing route line stop at the start point. Then it calls the `func updateRoute(_ routeProgress:)` to update the vanishing route line. 

To fix the problem, inside the `navigationMapView.show(_ routes: [Route], legIndex: Int = 0)`, it should replace the default `0.0` value with the `fractionTraveled`. And prepare a resonnable value of the fractionTraveled to be used.

Reply:
I create a new ticket to handle this problem  in #2757

Reply:
In the new commit, the refresh problem is solved with using the old `fractionTraveled` to form the two layers for vanishing route line. Then it receive updates from route progress, and slight change from the old `fractionTraveled` stop.